### PR TITLE
Fix console for test and migrate to work with public testnets and dashboard

### PR DESCRIPTION
**Pull request by sukanyaparashar**
### ISSUE
This issue was introduced by the merged PR #5245. `test` and `migrate` commands are not working for public testnets like `ropsten` or `goerli` with `truffle console --network <testnet or dashboard>`, which requires a `provider` property to connect to the testnet specified. This is because [the `provider` property of the network config is set here](https://github.com/trufflesuite/truffle/blob/54d1f3779e5bbf5d476f36a96b317f1cfee46183/packages/core/lib/console-child.js#L54), which doesn't have a check for the `provider` property's existence in a network config.

### SOLUTION
This PR adds a condition to check if `provider` is specified in a given network config and uses that `provider` to connect to the public testnet (`ropsten`, `goerli`...etc). It also checks if the network specified is `dashboard`.

### CODE REVIEW

**Review by cds-amal — COMMENTED** [review]
[Truffle docs](https://trufflesuite.com/docs/truffle/reference/configuration/#networks) show that there can be other optional values for a network. Should those be preserved, like the provider?

**Comment by sukanyaparashar** [comment]
> [Truffle docs](https://trufflesuite.com/docs/truffle/reference/configuration/#networks) show that there can be other optional values for a network. Should those be preserved, like the provider?

@amal, thanks for the comment. I guess we will need to have special consideration for the `dashboard` stuff since the rpc endpoint for `dashboard` is by default `http://localhost:24012/rpc`.

File: packages/core/lib/console-child.js
Comment (cds-amal) [inline packages/core/lib/console-child.js:65]:
```suggestion
  const isDashboardNetwork = network === "dashboard";
```
The result of the comparison is a Boolean

File: packages/core/lib/console-child.js
Comment (cds-amal) [inline packages/core/lib/console-child.js:41]: Is `configDefinesProvider` clearer than `configNetworkWithProvider`?

File: packages/core/lib/console-child.js
Comment (cds-amal) [inline packages/core/lib/console-child.js:45]: Should this be updated to indicate other configured network properties are used as well?

**Review by cds-amal — CHANGES_REQUESTED** [review]
Thanks, @sukanyaparashar, I left some minor fedback. Can you also add a couple of unit tests for this logic?

File: packages/core/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/lib/console-child.js:41]: Your suggestion makes sense. `configDefinesProvider` sounds like a better alternative.

File: packages/core/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/lib/console-child.js:45]: Yes, this comment needs to be updated I guess, to indicate spreading out of the other specified network properties. Thanks for this @cds-amal.

**Comment by sukanyaparashar** [comment]
> Thanks, @sukanyaparashar, I left some minor fedback. Can you also add a couple of unit tests for this logic?

I like your idea of adding tests for this logic, @amal. Do we need to add tests for the `dashboard` stuff too?

File: packages/core/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/lib/console-child.js]: Yeah! It should be!

File: packages/core/lib/command-utils.js
Comment (eggplantzzz) [inline packages/core/lib/command-utils.js:215]: Would it be slightly more flexible if the second argument were more general, like just `urlSuffix`? It probably does't matter too much in this case though.

File: packages/core/lib/command-utils.js
Comment (eggplantzzz) [inline packages/core/lib/command-utils.js:270]: Do we not want to spread after adding the defaults? That way user settings override the default ones.

File: packages/core/lib/command-utils.js
Comment (eggplantzzz) [inline packages/core/lib/command-utils.js:274]: I'm not sure it makes sense to specify both `host` and `url` along with `provider`. Shouldn't we only specify one or the other?

Also, if we spread the user's settings in here there may be a situation where we end up with both. In that case, what happens? Does one take precedence over the other?

File: packages/core/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/lib/console-child.js:50]: Why did you need to put all of the above to be executed in a function here at the end?

File: packages/core/test/sources/console-child/truffle-config.js
Comment (eggplantzzz) [inline packages/core/test/sources/console-child/truffle-config.js:10]: What a crrraaaaazy network!

File: packages/core/lib/command-utils.js
Comment (sukanyaparashar) [reply packages/core/lib/command-utils.js:270]: I think your suggestion makes sense @eggplantzzz. We can spread the `customConfig` after the defaults.

File: packages/core/lib/command-utils.js
Comment (sukanyaparashar) [reply packages/core/lib/command-utils.js:274]: In this case,`provider` takes precedence over the other. Because our logic says that if there is a `provider` specified in the config then we use the `provider` to connect to the network and ignore `host` and `port` (The first `if` block of the main function will be executed). Does that make sense?

File: packages/core/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/lib/console-child.js:50]: Oh! This is just for sometime in the future if we want this function to be exported for some reason.

File: packages/core/test/sources/console-child/truffle-config.js
Comment (sukanyaparashar) [reply packages/core/test/sources/console-child/truffle-config.js:10]: I guess it is not necessary. I will remove the `gas` property. Thanks @eggplantzzz.

File: packages/core/lib/console-child.js
Comment (cds-amal) [reply packages/core/lib/console-child.js:50]: IIRC, this came about from refactoring the module to make unit testing easier. Why would it be easier? It's because this module is executed as a child-process and those tests are more complicated; factoring the logic to a different module simplifies writing unit tests. 

It could be extended to use a `require.main === module ? main() : throw new Error("console-child should not be imported")` check to prevent incorrect import, but I don't think it's necessary.

File: packages/core/lib/command-utils.js
Comment (eggplantzzz) [reply packages/core/lib/command-utils.js:274]: Yes it does! But that seems to mean that we only want to add a provider here in the case that there is one since `host` and `port` will be ignored **every time** since a provider is always specified here, right?

Comment (sukanyaparashar) [reply packages/core/lib/command-utils.js:274]: That makes sense to me! We can remove the `host` and `port` settings as the provider will always be here. Thanks for catching this @eggplantzzz.

File: packages/core/lib/command-utils.js
Comment (cds-amal) [inline packages/core/lib/command-utils.js:238]: This line is not tested in the test, or did I miss it?

File: packages/core/test/lib/console-child.js
Comment (sukanyaparashar) [inline packages/core/test/lib/console-child.js:37]: Does this line satisfies your query @cds-amal ? If the `crazyTimeNetwork` has other properties, then it should be included in the `cfg.networks.crazyTimeNetwork`. Please correct me if I am wrong somewhere :)

Comment (sukanyaparashar) [reply packages/core/test/lib/console-child.js]: Hey @amal. Please ignore the above comment. I realised that there were some mistakes in the unit tests. I have updated the logic for each of the use cases.

Comment (sukanyaparashar) [reply packages/core/test/lib/console-child.js]: Hey @cds-amal. I realised that there were a few mistakes in the unit tests. I have updated the logic. Please ignore my above comment.

File: packages/core/lib/command-utils.js
Comment (cds-amal) [inline packages/core/lib/command-utils.js:239]: Is this necessary if the entire network object, which should include the provider property, is spread out?

**Review by cds-amal — COMMENTED** [review]
The tests look better! I will have to take a look tomorrow.

File: packages/core/lib/command-utils.js
Comment (sukanyaparashar) [reply packages/core/lib/command-utils.js:239]: Yeah! This is unnecessary as the network object will include the `provider` specified. I will remove this line. Thanks @cds-amal.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:32]: These strings should read as plain english descriptions of what these things do. An example is something like
```
it("returns a config with...<whatever you expect>...
```

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:25]: Get rid of "Create a sandbox", it is obvious from `createSandbox` that that is what's happening :)

Just omit the string like `before(function () {...`

File: packages/core/test/sources/console-child/truffle-config.js
Comment (eggplantzzz) [reply packages/core/test/sources/console-child/truffle-config.js:9]: Oh, I guess that stuff would be useful if you wanted to test that user-specified parameters end up in the return for this function.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:35]: Here you should probably just omit the third argument or explicitly pass `undefined` as I think that will most closely model the situation where `url` is not present. In other words, if no `url` is present, an empty string will not be passed but rather `undefined`.

Also a small stylistic nit; I would kind of rather specify the string literal `"crazyTimeNetwork"` as the second argument and not use `config.network`. I think it is slightly clearer and I'd rather be clearer than use another value that just happens to work.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:47]: Same here about using a string literal for the network name.

File: packages/core/test/sources/console-child/truffle-config.js
Comment (cds-amal) [reply packages/core/test/sources/console-child/truffle-config.js:9]: ah, this is related to my comment as well. You could stick a `customUserProperty` in the network config and assert it shows up in the verification step.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:58]: Same here about the string literal and `undefined` for the `url` parameter.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:44]: I kind of feel like you should change this network name so it doesn't match the url provivded as a third argument. It seems it is likely to confuse people in that they might think there is some relationship between the third argument provided and this network name when, in fact, they are totally unrelated.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:33]: In all these cases I would prefer that you not mutate `config` and instead use string literals as arguments as I've mentioned in a couple of places.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:59]: Just use a string literal instead of `config.network`.

**Comment by eggplantzzz** [comment]
In the tests for this, I think `config` should just remain an unmutated object. Just use the config as input to the function that you are testing and instead of setting `config.network` to something different for each test (this is unnecessary and confusing I think), just use a string with the network name.

File: packages/core/test/sources/console-child/truffle-config.js
Comment (eggplantzzz) [inline packages/core/test/sources/console-child/truffle-config.js:11]: This should be a provider if it is set as `provider` and not a string.

File: packages/core/test/lib/console-child.js
Comment (eggplantzzz) [inline packages/core/test/lib/console-child.js:75]: Same with this test with just using the network name and not setting `config.network`.

File: packages/core/test/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/test/lib/console-child.js:32]: I will change the strings according to your suggestions @eggplantzzz. Thanks !

File: packages/core/test/sources/console-child/truffle-config.js
Comment (sukanyaparashar) [reply packages/core/test/sources/console-child/truffle-config.js:9]: Yeah! I agree with you @cds-amal. I will add a `customUserProperty` in the `network` config. Thanks for your suggestions @cds-amal and @eggplantzzz.

File: packages/core/test/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/test/lib/console-child.js:44]: When we specify the `--url` option with `truffle console` command, then the `network` name happens to be the `host` of the url. It is the logic that goes behind `mergeConfigNetwork.js` inside the `core/lib` package.

Comment (eggplantzzz) [reply packages/core/test/lib/console-child.js:44]: Yeah I understand, but the network name in this case is arbitrary and has no bearing on the results (at least in this function). In other words, you could name it `"sukanyaNetwork"` and this test will correctly pass. I'm suggesting you change it to something like that so people won't think it has something to do with the `url` parameter.

**Comment by sukanyaparashar** [comment]
> In the tests for this, I think `config` should just remain an unmutated object. Just use the config as input to the function that you are testing and instead of setting `config.network` to something different for each test (this is unnecessary and confusing I think), just use a string with the network name.

Okay! That makes sense to me. Thanks @eggplantzzz.

File: packages/core/test/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/test/lib/console-child.js:44]: I got your point @eggplantzzz. The network name doesn't matter here much. I will apply your suggestions :)

File: packages/core/test/lib/command-utils.js
Comment (cds-amal) [inline packages/core/test/lib/command-utils.js:53]: This value is used 2x, it can be factored out as a constant.

**Review by cds-amal — CHANGES_REQUESTED** [review]
Thanks, @sukanyaparashar! Aren't unit tests great? I only have a request to document a few functions, and the other comments are stylistic and can be ignored.

File: packages/core/lib/console-child.js
Comment (sukanyaparashar) [reply packages/core/lib/console-child.js:17]: I guess we can go with the name `deriveConfig` then?

File: packages/core/test/lib/command-utils.js
Comment (sukanyaparashar) [reply packages/core/test/lib/command-utils.js:53]: Yeah! I missed that. Thanks @cds-amal.

**Comment by sukanyaparashar** [comment]
> Thanks, @sukanyaparashar! Aren't unit tests great? I only have a request to document a few functions, and the other comments are stylistic and can be ignored.

I agree with you @cds-amal. Unit tests are really great ! :)

File: packages/core/lib/console-child.js
Comment (cds-amal) [reply packages/core/lib/console-child.js]: This was a stylistic comment, feel free to ignore it.

Comment (sukanyaparashar) [reply packages/core/lib/console-child.js]: I thought `deriveConfig` will be a better name, so I changed it in my recent commit. :)

**Review by eggplantzzz — APPROVED** [review]
Looks great! I guess @cds-amal has something else to be addressed? I am good with this.

**Comment by sukanyaparashar** [comment]
> Looks great! I guess @cds-amal has something else to be addressed? I am good with this.

Thanks for the approval @eggplantzzz. I will wait for @cds-amal's comment and approval as well. :)

**Review by cds-amal — APPROVED** [review]
Looks good!

**Pull request merged**
